mlflow/system_metrics/metrics/gpu_monitor.py

```diff
@@ -0,0 +1,48 @@
+"""Class for monitoring GPU stats."""
+
+import logging
+import sys
+
+from mlflow.system_metrics.metrics.base_metrics_monitor import BaseMetricsMonitor
+
+_logger = logging.getLogger(__name__)
+
+try:
+    import pynvml
+except ImportError:
+    _logger.warning(
+        "`GPUMonitor` requires pynvml package. To install, run `pip install pynvml`. Skip "
+        "logging GPU metrics."
+    )
+
+
+class GPUMonitor(BaseMetricsMonitor):
+    """Class for monitoring GPU stats."""
+
+    def __new__(cls, *args, **kwargs):
+        # Override `__new__` to return a None object if `pynvml` is not installed or GPU is not
+        # found.
+        if "pynvml" not in sys.modules:
+            # Only instantiate if `pynvml` is installed.
+            return
+        try:
+            # `nvmlInit()` will fail if no GPU is found.
+            pynvml.nvmlInit()
+        except pynvml.NVMLError as e:
+            _logger.warning(f"Failed to initialize NVML, skip logging GPU metrics: {e}")
+            return
+        return super().__new__(cls, *args, **kwargs)
+
+    def __init__(self, name="gpu"):
+        super().__init__(name)
+        self.num_gpus = pynvml.nvmlDeviceGetCount()
+        self.gpu_handles = [pynvml.nvmlDeviceGetHandleByIndex(i) for i in range(self.num_gpus)]
+
+    def collect_metrics(self):
+        # Get GPU metrics.
+        for i, handle in enumerate(self.gpu_handles):
```

Comment:
this doesn't handle auto scaling systems, right? Can we refresh the gpu_handles index values and persist the enumeration indexes based on a rubric each trigger window for logging? 
Is the overhead for this operation excessive if we allow for adapting for auto scaling? 
What happens if a GPU instance is removed from a long-duration training run and the handle ID is no longer available for nvml to monitor? Will that abort the training run or result in a loss of metric collection data?

Reply:
good question. Yes it does not handle auto-scaling. 

My understanding is GPU auto-scaling is only useful for inference. During training, adding a new GPU requires restarting the training process, so the system metrics logging will be restarted as well.

Reply:
I'm more curious about how distributed auto-recovery training will work with usage monitoring. A lot of GPU training is done with RayTune these days and it has fault tolerance (it will resume from last batch epoch without user intervention) https://docs.ray.io/en/latest/train/dl_guide.html#fault-tolerance . They're also working on elastic training support, so this will be a thing very soon.

Reply:
thanks for the context! I read doc the to my understanding, this is handling data parallelism, so each worker receives the full training code binary and partial data. In this scenario, when a new worker is added, it will start its own job, and also a new system metrics logging thread, so theoretically it should just work. One thing we may want to take care of is how to add identifier to these workers. Not necessarily with this PR, but definitely worth discussion!

For model/tensor parallelism, if a worker crashes I think the whole job needs to restart from the latest checkpoint, because losing part of the weights could cause bad results on the training. So in this scenario,  system metrics logging might not need a special treatment.

Reply:
@chenmoneygithub Can you test what happens if we autoscale on a Databricks cluster with GPUs? Can we actually query info about each GPU from the driver? I suspect not, which is totally fine.

Reply:
Yes, I am asking around for multi-gpu access, and I think I can use Mosaic cluster to test. Will loop back, probably non-blocking tho

Reply:
Btw, just to clarify, all of the metrics logging in this PR only covers single node (rank 0), right? So none of these metrics will work for multiple nodes (or am I missing something here)?

Reply:
multi-node is a complex setup, and it will require extra code. I don't think Mlflow can provide a one-for-all solution for multi-node logging, because of different platforms have different ways to configure multi-node cluster. For reference, wand requires extra steps on multi-node logging: https://docs.wandb.ai/guides/track/log/distributed-training

We may not want to handle multi-gpu with this PR, which might be more complex than everything we get here. I will follow up with this, working on a multi-gpu cluster setup now.